src/FSharp.Control.TaskSeq/TaskSeqInternal.fs

```diff
@@ -732,134 +732,85 @@ module internal TaskSeqInternal =
-            let! moveFirst = e.MoveNextAsync()
-            let mutable more = moveFirst
-
-            match whileKind, predicate with
-            | Exclusive, Predicate predicate -> // skipWhile
-                while more && predicate e.Current do
-                    let! hasMore = e.MoveNextAsync()
-                    more <- hasMore
-
-                if more then
-                    // yield the last one where the predicate was false
-                    // (this ensures we skip 0 or more)
-                    yield e.Current
-
-                    while! e.MoveNextAsync() do // get the rest
-                        yield e.Current
-
-            | Inclusive, Predicate predicate -> // skipWhileInclusive
-                while more && predicate e.Current do
-                    let! hasMore = e.MoveNextAsync()
-                    more <- hasMore
-
-                if more then
-                    // yield the rest (this ensures we skip 1 or more)
-                    while! e.MoveNextAsync() do
-                        yield e.Current
-
-            | Exclusive, PredicateAsync predicate -> // skipWhileAsync
-                let mutable cont = true
-
-                if more then
-                    let! hasMore = predicate e.Current
-                    cont <- hasMore
-
-                while more && cont do
-                    let! moveNext = e.MoveNextAsync()
-
-                    if moveNext then
-                        let! hasMore = predicate e.Current
-                        cont <- hasMore
 
-                    more <- moveNext
-
-                if more then
-                    // yield the last one where the predicate was false
-                    // (this ensures we skip 0 or more)
-                    yield e.Current
+            match! e.MoveNextAsync() with
+            | false -> () // Nothing further to do, no matter what the rules are
+            | true ->
 
-                    while! e.MoveNextAsync() do // get the rest
-                        yield e.Current
+                let exclusive =
+                    match whileKind with
+                    | Exclusive -> true
+                    | Inclusive -> false
 
-            | Inclusive, PredicateAsync predicate -> // skipWhileInclusiveAsync
                 let mutable cont = true
 
-                if more then
-                    let! hasMore = predicate e.Current
-                    cont <- hasMore
+                match predicate with
+                | Predicate predicate -> // skipWhile(Inclusive)?
+                    while cont do
+                        if predicate e.Current then // spam -> skip
+                            let! hasAnother = e.MoveNextAsync()
+                            cont <- hasAnother
+                        else // Starting the ham
+                            if exclusive then
+                                yield e.Current // return the item as it does not meet the condition for skipping
```

Comment:
It is rather confusing to call something `exclusive` and then return a value. It is probably better to call it `skipInclusive`, to be closer to the meaning of the actual function you are implementing and one less `not not` inverse jump in your mind to take.

Don't change it yet, I have to get an idea if I find this mixing of functions inside single bodies a good idea. I kind like it is the compiled function that people are using is only one of the function bodies, and not a combination of two or more (but this is no hard and fast rule, though there are obvious perf implications involved).